### PR TITLE
:tada: Adds a Multithreading Tip Article

### PR DESCRIPTION
**Pull request by AlexandreSajus**
Adds a Tip Article about displaying in Taipy data sent from another thread. It uses a simple example of sending a random number through a socket and displaying it. This tutorial has been requested by a user on Discord.

image

This is my first PR on taipy-doc, so let me know what changes I need to make. For example, I have not added this article to the list of tip articles.

### CODE REVIEW

File: docs/knowledge_base/tips/multithreading/index.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
+Taipy can display data that is generated in a separate thread. This is useful for displaying real-time data from a sensor or a simulation.
+
+<p align="center">
+  <img src="realtime_dashboard.png" alt="Dashboard Example" width="80%"/>
+</p>
+
+In this article, we will code a simple example where:
+- A sender.py script will generate a random number and send it through a socket.
```
Comment (FabienLelaquais) [inline docs/knowledge_base/tips/multithreading/index.md]: Please use code style when talking about file names: \`sender.py\` (and \`receiver.py\` below)

File: docs/knowledge_base/tips/multithreading/index.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+- A receiver.py script will receive and display the number in a Taipy app.
```
Comment (FabienLelaquais) [inline docs/knowledge_base/tips/multithreading/index.md]: app -> application

File: docs/knowledge_base/tips/multithreading/index.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+<p align="center">
+  <img src="vscode_screen.png" alt="VSCode Screenshot" width="80%"/>
+</p>
+
+## Step 1: Create the Sender Script
+
+Here is the code for the sender.py script:
```
Comment (FabienLelaquais) [inline docs/knowledge_base/tips/multithreading/index.md]: Backticks

File: docs/knowledge_base/tips/multithreading/index.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+```python
```
Comment (FabienLelaquais) [inline docs/knowledge_base/tips/multithreading/index.md]: Add `title="sender.py"`

File: docs/knowledge_base/tips/multithreading/index.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+    s.connect((HOST, PORT))
+    while True:
+        random_number = randint(1, 100)
+        s.sendall(str(random_number).encode())
+        print(f"Sending: {random_number}")
+        time.sleep(5)
+```
+
+This script will generate a random number between 1 and 100, send it through a socket, and wait 5 seconds before sending another number.
```
Comment (FabienLelaquais) [inline docs/knowledge_base/tips/multithreading/index.md]: "will generate" -> "generates"
"send" -> "sends"
"wait" -> "waits"

File: docs/knowledge_base/tips/multithreading/index.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+## Step 2: Create the Receiver Script
+
+Coding the receiver script requires multiple steps:
+
+1. Imports and defining the socket parameters.
+
+```python
```
Comment (FabienLelaquais) [inline docs/knowledge_base/tips/multithreading/index.md]: Add `title="receiver.py"`

File: docs/knowledge_base/tips/multithreading/index.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+from threading import Thread
+from taipy.gui import Gui, State, invoke_callback, get_state_id
+
+HOST = "127.0.0.1"
+PORT = 5050
+```
+
+2. We gather the list of state_ids. These are identifiers of the clients connected to our Taipy app. We need this list to choose which client to send the data to.
+
+```python
```
Comment (FabienLelaquais) [inline docs/knowledge_base/tips/multithreading/index.md]: Add `title="receiver.py"`

File: docs/knowledge_base/tips/multithreading/index.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+import socket
+from threading import Thread
+from taipy.gui import Gui, State, invoke_callback, get_state_id
+
+HOST = "127.0.0.1"
+PORT = 5050
+```
+
+2. We gather the list of state_ids. These are identifiers of the clients connected to our Taipy app. We need this list to choose which client to send the data to.
```
Comment (FabienLelaquais) [inline docs/knowledge_base/tips/multithreading/index.md]: "list of state_ids" -> "list of state identifiers"
"app" -> "application"

File: docs/knowledge_base/tips/multithreading/index.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+def on_init(state: State):
+    state_id = get_state_id(state)
+    if (state_id := get_state_id(state)) is not None and state_id != "":
+        state_id_list.append(state_id)
+```
+
+3. We create a function to listen to the socket. When the socket receives data, it triggers a callback to send the data to the Taipy app of one of the connected clients.
+
+```python
```
Comment (FabienLelaquais) [inline docs/knowledge_base/tips/multithreading/index.md]: Add `title="receiver.py"`

File: docs/knowledge_base/tips/multithreading/index.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+```
+
+2. We gather the list of state_ids. These are identifiers of the clients connected to our Taipy app. We need this list to choose which client to send the data to.
+
+```python
+state_id_list = []
+
+def on_init(state: State):
+    state_id = get_state_id(state)
+    if (state_id := get_state_id(state)) is not None and state_id != "":
```
Comment (FabienLelaquais) [inline docs/knowledge_base/tips/multithreading/index.md]: I'm not sure that `and state_id != ""` makes sense...

File: docs/knowledge_base/tips/multithreading/index.md [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+        state_id_list.append(state_id)
+```
+
+3. We create a function to listen to the socket. When the socket receives data, it triggers a callback to send the data to the Taipy app of one of the connected clients.
```
Comment (FabienLelaquais) [inline docs/knowledge_base/tips/multithreading/index.md]: app -> application
"of one" -> "for one"

File: docs/knowledge_base/tips/multithreading/index.md [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        else:
+            print("Connection closed")
+            break
+
+
+def update_received_data(state: State, val):
+    state.received_data = val
+```
+
+4. We create the Taipy app to display the data. We run both the client_handler and the Taipy app in separate threads.
```
Comment (FabienLelaquais) [inline docs/knowledge_base/tips/multithreading/index.md]: app -> application (twice)
"We run both the client_handler and the Taipy app in separate threads." -> "The \*client_handler()\* function and the application itself are run in different threads.'

File: docs/knowledge_base/tips/multithreading/index.md [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+
+```python
```
Comment (FabienLelaquais) [inline docs/knowledge_base/tips/multithreading/index.md]: Add `title="receiver.py"`

File: docs/knowledge_base/tips/multithreading/index.md [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    ),
+)
+t.start()
+
+gui.run(run_browser=False)
+```
+
+## Step 3: Run the Scripts
+
+Run the receiver.py script first, then the sender.py script in another terminal. The receiver will receive and display the sender's data in the Taipy app.
```
Comment (FabienLelaquais) [inline docs/knowledge_base/tips/multithreading/index.md]: Backticks on filenames.
app -> application

**Review by FabienLelaquais — APPROVED** [review]
Real nice!
A few typographic changes

File: docs/knowledge_base/tips/multithreading/index.md [hunk 15]
```diff
@@ -0,0 +1,113 @@
+Taipy can display data that is generated in a separate thread. This is useful for displaying real-time data from a sensor or a simulation.
```
Comment (jrobinAV) [inline docs/knowledge_base/tips/multithreading/index.md]:
```suggestion
Taipy can display data that is generated in a separate thread. This is useful for displaying real-time data from a sensor or a simulator.
```

File: docs/knowledge_base/tips/multithreading/index.md [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
```
Comment (jrobinAV) [inline docs/knowledge_base/tips/multithreading/index.md:5]: What do you think about saying that it is very useful for real-time monitoring of continuously updated data about systems, processes, or events? And maybe providing a few examples like real-time CPU/memory consumption of an application or a machine, manufacturing line efficiency, financial transactions, etc. ?

**Review by jrobinAV — COMMENTED** [review]
Great PR ! Not much to say on the content.

Please update the following files as well: 
- `taipy-doc/docs/knowloedge_base/index.md` to list your tip in the knowledge_base page. You need to provide a title and a description for your tip. You can take another tip as reference to see how to do it. Since you don't have any video on it, please use the icon-code.svg like other the tips.
- `taipy-doc/docs/knowloedge_base/tips/index.md` to list your tip in the tips index page. You need to provide a title, a description and an image for your tip. I believe you can reuse the `realtime_dashboard.png` pic that is good.
- `taipy-doc/mkdocs.yml_template` to make your page appear in the navigation panel. You need to provide a title for the navigation panel.

**Review by FlorianJacta — APPROVED** [review]

File: docs/knowledge_base/tips/multithreading/index.md (hunk 4 above)
Comment (AlexandreSajus) [reply docs/knowledge_base/tips/multithreading/index.md:19]: What do you mean?

Comment (jrobinAV) [reply docs/knowledge_base/tips/multithreading/index.md]:
```suggestion
```python title="sender.py"
```

File: docs/knowledge_base/tips/multithreading/index.md (hunk 6 above)
Comment (AlexandreSajus) [reply docs/knowledge_base/tips/multithreading/index.md]: I added this but what does this do?

File: docs/knowledge_base/tips/multithreading/index.md (hunk 16 above)
Comment (AlexandreSajus) [reply docs/knowledge_base/tips/multithreading/index.md:5]: The use case I created this tutorial for initially was a company that works in IoT with sensors measuring metrics in a field. They asked us if Taipy could create a dashboard that displayed real-time data from these sensors. I have added some more context in the next commit. Feel free to add more.

**Comment by AlexandreSajus** [comment]
I implemented the requested changes. Please check that the navigation works well. I have not built the doc yet.

File: docs/knowledge_base/tips/multithreading/index.md [hunk 17]
```diff
@@ -0,0 +1,113 @@
+Taipy can display data that is generated in a separate thread. This is useful for displaying real-time data from a sensor or a simulator. For example displaying in a dashboard the information from sensors measuring air pollution around a city, or displaying CPU usage of a server.
```
Comment (jrobinAV) [inline docs/knowledge_base/tips/multithreading/index.md]: In order to facilitate reviews in GitHub, it is better to have lines with max 100 chars. 
At least for text content. It's not always feasible on code blocks, links, or html tags.

```suggestion
Taipy can display data that is generated in a separate thread. This is useful for displaying
real-time data from a sensor or a simulator. For example displaying in a dashboard the
information from sensors measuring air pollution around a city, or displaying CPU usage of
a server.
```

File: docs/knowledge_base/tips/multithreading/index.md [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+    s.connect((HOST, PORT))
+    while True:
+        random_number = randint(1, 100)
+        s.sendall(str(random_number).encode())
+        print(f"Sending: {random_number}")
+        time.sleep(5)
+```
+
+This script generates a random number between 1 and 100, sends it through a socket, and waits 5 seconds before sending another number.
```
Comment (jrobinAV) [inline docs/knowledge_base/tips/multithreading/index.md]:
```suggestion
This script generates a random number between 1 and 100, sends it through a socket, and waits
5 seconds before sending another number.
```

File: docs/knowledge_base/tips/multithreading/index.md [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+```python title="receiver.py"
+import socket
+from threading import Thread
+from taipy.gui import Gui, State, invoke_callback, get_state_id
+
+HOST = "127.0.0.1"
+PORT = 5050
+```
+
+2. We gather the list of state identifiers. These are identifiers of the clients connected to our Taipy application. We need this list to choose which client to send the data to.
```
Comment (jrobinAV) [inline docs/knowledge_base/tips/multithreading/index.md]:
```suggestion
2. We gather the list of state identifiers. These are identifiers of the clients connected to
our Taipy application. We need this list to choose which client to send the data to.
```

File: docs/knowledge_base/tips/multithreading/index.md [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+```python title="receiver.py"
+state_id_list = []
+
+def on_init(state: State):
+    state_id = get_state_id(state)
+    if (state_id := get_state_id(state)) is not None:
+        state_id_list.append(state_id)
+```
+
+3. We create a function to listen to the socket. When the socket receives data, it triggers a callback to send the data to the Taipy application for one of the connected clients.
```
Comment (jrobinAV) [inline docs/knowledge_base/tips/multithreading/index.md]:
```suggestion
3. We create a function to listen to the socket. When the socket receives data, it triggers a
callback to send the data to the Taipy application for one of the connected clients.
```

File: docs/knowledge_base/tips/multithreading/index.md [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+        else:
+            print("Connection closed")
+            break
+
+
+def update_received_data(state: State, val):
+    state.received_data = val
+```
+
+4. We create the Taipy application to display the data. The *client_handler()* function and the application itself are run in different threads.
```
Comment (jrobinAV) [inline docs/knowledge_base/tips/multithreading/index.md]:
```suggestion
4. We create the Taipy application to display the data. The *client_handler()* function and
the application itself are run in different threads.
```

File: docs/knowledge_base/tips/multithreading/index.md [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+    ),
+)
+t.start()
+
+gui.run(title="Receiver Page")
+```
+
+## Step 3: Run the Scripts
+
+Run the `receiver.py` script first, then the `sender.py` script in another terminal. The receiver will receive and display the sender's data in the Taipy application.
```
Comment (jrobinAV) [inline docs/knowledge_base/tips/multithreading/index.md]:
```suggestion
Run the `receiver.py` script first, then the `sender.py` script in another terminal. The
receiver will receive and display the sender's data in the Taipy application.
```

File: mkdocs.yml_template [hunk 23]
```diff
@@ -69,6 +69,7 @@ nav:
         - "Skippable Tasks": knowledge_base/tips/skippable_tasks/index.md
         - "Callbacks in Taipy": knowledge_base/tips/the_on_change_callback/index.md
         - "Taipy Cloud": knowledge_base/tips/taipy_cloud_deploy/index.md
+        - "Multithreading": knowledge_base/tips/multithreading/index.md
```
Comment (jrobinAV) [inline mkdocs.yml_template]: This is the title exposed to the reader. I believe we want something more functional than technical.

```suggestion
        - "Real-time data visualization": knowledge_base/tips/multithreading/index.md
```

Comment (jrobinAV) [reply mkdocs.yml_template]: @AlexandreSajus Are you ok with this new title?

**Pull request merged**